src/Compilers/Core/Portable/Text/StringBuilderText.cs

```diff
@@ -89,5 +91,47 @@ public override void CopyTo(int sourceIndex, char[] destination, int destination
         {
             _builder.CopyTo(sourceIndex, destination, destinationIndex, count);
         }
+
+        public override void Write(TextWriter textWriter, TextSpan span, CancellationToken cancellationToken = default)
+        {
+            CheckSubSpan(span);
+
+            if (span.Length == 0)
+                return;
```

Comment:
These checks need to be repeated in each override, which isn't ideal. I thought about just having these in the base Write method and having it forward to a new WriteCore method that is actually virtual and overriden, but:
1. Write is a public method so making it non-virtual would be an API change. I guess this class can only be derived internally though so it could be OK?
2. With that change, if you have a derived class and call Write on it, it would now go through virtual dispatch, whereas now, it doesn't have to if the class is sealed.

Reply:
> I guess this class can only be derived internally though so it could be OK?

It can be really difficult to tell this by looking at the code. The rules for what we consider truly public vs. not are easy to mess up. The easiest way to tell though is look at `PublicAPI.Shipped.txt`. If the API is there then it's public and 3rd parties can derive from it. In this case it is public. 

https://github.com/dotnet/roslyn/blob/main/src/Compilers/Core/Portable/PublicAPI.Shipped.txt#L3957

Reply:
@jaredpar It is public, but SourceText has internal virtual methods, which means that external code cannot derive from it, right?

Reply:
This is not easy to tell at a glance. I spent a good 5 minutes searching around before I wrote my previous reply because I couldn't remember if `SourceText` was derivable or not.

Reply:
is it just this first test?  if so, it seems fine to repeat.  the code will be extremely clear and hard to mess up.

Reply:
@CyrusNajmabadi Yes, it's clear like this, but it's easy to forget to do this when someone overrides the method - eg. in `StringText`, the bounds checking was missing before this PR.